src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -7267,10 +7267,7 @@ private void SaveError(ParseError error)
 
         private void SaveError(IScriptExtent extent, string errorId, string errorMsg, bool incompleteInput, params object[] args)
         {
-            if (String.IsNullOrEmpty(errorId))
-            {
-                errorId = "ParserError";
-            }
+            Diagnostics.Assert(!String.IsNullOrEmpty(errorId), "Errors must have IDs");
```

Comment:
It might even be worth an assertion that the errorMsg is exactly ParserString.errorId, so something like (not tested, probably doesn't even compile):

```c#
Diagnostics.Assert(typeof(ParserStrings).GetProperty(errorId, null).Value.Equals(errorMsg), "ErrorId must match errorMsg from ParserStrings);
```

Reply:
I'd consider moving the resources instead of writing a helper function.

Assertions disappear in non-debug builds - the arguments should not be evaluated - but to be honest, I've never tried a non-trivial argument like what I wrote - I'm just assuming the jit removes the argument evaluation.

Reply:
There are also a couple of hand-rolled strings passed into `SaveError`. The ones I saw were "this should never happen" clauses. Perhaps I should just migrate everything to ParserStrings.resx and include a new "ParserError", "{0} is very bad and should never ever happen" entry?